### PR TITLE
Support double precision in LogisticRegression

### PR DESCRIPTION
**Pull request by lijinf2**

### CODE REVIEW

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/classification.py [hunk 1]
```diff
@@ -860,6 +860,8 @@ class LogisticRegression(
         If False, always uses dense array. This is favorable if the majority of VectorUDT vectors are DenseVector.
         If True, always uses sparse array. This is favorable if the majority of the VectorUDT vectors are SparseVector.
         Note this is only supported in spark >= 3.4.
+    float32_inputs: bool (default=True):
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/classification.py:863]: We have opted to not add this as a documented constructor argument.   And document only here:  https://nvidia.github.io/spark-rapids-ml/compatibility.html#single-vs-double-precision-inputs  
Should we revisit that?

File: python/src/spark_rapids_ml/core.py [hunk 2]
```diff
@@ -176,13 +176,13 @@ def _get_unwrapped_vec_cols(feature_col: Column) -> List[Column]:
...
+
+    feature_type = FloatType() if float32_input is True else DoubleType()
     return [
         features_unwrapped_vec_col.type.alias(alias.featureVectorType),
         features_unwrapped_vec_col.size.alias(alias.featureVectorSize),
         features_unwrapped_vec_col.indices.alias(alias.featureVectorIndices),
-        # Note: the value field is double array type, cast it to float32 array type
-        # for speedup following repartitioning.
-        features_unwrapped_vec_col.values.cast(ArrayType(FloatType())).alias(
+        features_unwrapped_vec_col.values.cast(ArrayType(feature_type)).alias(
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/core.py:185]: If float32_input = true can we avoid the redundant cast to double?

File: python/tests/test_logistic_regression.py [hunk 3]
```diff
@@ -1986,3 +2021,84 @@ def sparse_to_df(X: csr_matrix, y: List[float]) -> DataFrame:
...
+    fit_intercept: bool,
+    reg_factors: Tuple[float, float],
+    feature_type: str,
+    data_shape: Tuple[int, int],
+    max_record_batch: int,
+    n_classes: int,
+    gpu_number: int,
+) -> None:
+
+    random_bool = random.choice([True, False])
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:2045]: Would be good to make these deterministic via seed for reproducibility.

File: python/tests/test_logistic_regression.py [hunk 4]
```diff
@@ -998,8 +1008,23 @@ def test_multiclass(
         gpu_number=gpu_number,
         reg_param=0.1,
         tolerance=tolerance,
+        set_float32_inputs=True,
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:1011]: set_float32_inputs=(data_type == np.float32) or similar would be simpler logic than repeating the whole instantiation.

File: python/tests/test_logistic_regression.py [hunk 5]
```diff
@@ -305,9 +299,16 @@ def test_classifier(
...
     tolerance: float = 0.001,
     convert_to_sparse: bool = False,
+    set_float32_inputs: Optional[bool] = None,
 ) -> LogisticRegression:
     standardization: bool = False
 
+    float32_inputs = set_float32_inputs
+    if float32_inputs is None:
+        float32_inputs = True
+        if data_type == np.float64 and max_record_batch > 1000:
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:309]: Add a comment explaining reason for this logic.

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/classification.py (hunk 1 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/classification.py]: Did not notice that. Revised to be consistent with compatibility.html.

File: python/src/spark_rapids_ml/core.py (hunk 2 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py]: Good catch!

File: python/tests/test_logistic_regression.py (hunk 3 above)
Comment (lijinf2) [reply python/tests/test_logistic_regression.py:2030]: Done.

File: python/tests/test_logistic_regression.py (hunk 4 above)
Comment (lijinf2) [reply python/tests/test_logistic_regression.py]: Need to test the case converting float64 to float32. 
Revised to use randomized value.

File: python/tests/test_logistic_regression.py (hunk 5 above)
Comment (lijinf2) [reply python/tests/test_logistic_regression.py]: Revised the use randomized value.

**Review by eordentlich — APPROVED** [review]
👍

**Pull request merged**